File: src-tauri/src/tauri_commands.rs
use crate::app::{ApplicationManager, TestType};
use crate::test_trial::TrialStatus;

use log::info;
use std::sync::Mutex;
use tauri::State;

#[tauri::command(rename_all = "snake_case")]
pub fn add_test(
    app_manager: State<Mutex<ApplicationManager>>,
    test_type: TestType,
    json_string: String,
) -> Result<(), String> {
    info!("add_test command is invoked");
    let result = app_manager.lock().unwrap().add_test(test_type, json_string);
    match result {
        Ok(_) => return Ok(()),
        Err(s) => return Err(s.to_string()),
    }
}

#[tauri::command(rename_all = "snake_case")]
pub fn delete_test(
    app_manager: State<Mutex<ApplicationManager>>,
    test_name: String,
) -> Result<(), String> {
    info!("delete_test command is invoked");
    let result = app_manager.lock().unwrap().delete_test(test_name);
    match result {
        Ok(_) => return Ok(()),
        Err(s) => return Err(s.to_string()),
    }
}

#[tauri::command(rename_all = "snake_case")]
pub fn edit_test(
    app_manager: State<Mutex<ApplicationManager>>,
    test_name: String,
    json_string: String,
) -> Result<(), String> {
    info!("edit_test command is invoked");
    let result = app_manager
        .lock()
        .unwrap()
        .edit_test(test_name, json_string);
    match result {
        Ok(_) => return Ok(()),
        Err(s) => return Err(s.to_string()),
    }
}

#[tauri::command(rename_all = "snake_case")]
pub fn start_test(
    app_manager: State<Mutex<ApplicationManager>>,
    test_name: String,
    examinee: String,
) -> Result<(), String> {
    info!("start_test command is invoked");
    let result = app_manager.lock().unwrap().start_test(test_name, examinee);
    match result {
        Ok(_) => return Ok(()),
        Err(s) => return Err(s.to_string()),
    }
}

#[tauri::command(rename_all = "snake_case")]
pub fn close_test(
    app_manager: State<Mutex<ApplicationManager>>,
    examinee: String,
) -> Result<(), String> {
    info!("close_test command is invoked");
    let result = app_manager.lock().unwrap().close_test(examinee);
    match result {
        Ok(_) => return Ok(()),
        Err(s) => return Err(s.to_string()),
    }
}

#[tauri::command(rename_all = "snake_case")]
pub fn get_audio(app_manager: State<Mutex<ApplicationManager>>) -> Result<Vec<String>, String> {
    info!("get_audio command is invoked");
    let result = app_manager.lock().unwrap().get_audio();
    match result {
        Ok(p) => {
            let paths: Vec<String> = p
                .into_iter()
                .map(|_p| _p.to_string_lossy().into_owned())
                .collect();
            return Ok(paths);
        }
        Err(s) => return Err(s.to_string()),
    }
}

#[tauri::command(rename_all = "snake_case")]
pub fn set_score(
    app_manager: State<Mutex<ApplicationManager>>,
    score: Vec<String>,
) -> Result<TrialStatus, String> {
    info!("set_score command is invoked");
    let result = app_manager.lock().unwrap().set_score(score);
    match result {
        Ok(v) => return Ok(v),
        Err(s) => return Err(s.to_string()),
    }
}

#[tauri::command(rename_all = "snake_case")]
pub fn get_settings(app_manager: State<Mutex<ApplicationManager>>) -> Result<Vec<String>, String> {
    info!("get_setting command is invoked");
    let result = app_manager.lock().unwrap().get_settings();
    match result {
        Ok(v) => return Ok(v),
        Err(s) => return Err(s.to_string()),
    }
}

#[tauri::command(rename_all = "snake_case")]
pub fn delete_trial(
    app_manager: State<Mutex<ApplicationManager>>,
    test_name: String,
    examinee: String,
) -> Result<(), String> {
    info!("delete_trial command is invoked");
    let result = app_manager
        .lock()
        .unwrap()
        .delete_trial(test_name, examinee);
    match result {
        Ok(_) => return Ok(()),
        Err(s) => return Err(s.to_string()),
    }
}

#[tauri::command(rename_all = "snake_case")]
pub fn start_preview(
    app_manager: State<Mutex<ApplicationManager>>,
    test_name: String,
) -> Result<(), String> {
    info!("start_preview command is invoked");
    let result = app_manager.lock().unwrap().start_preview(test_name);
    match result {
        Ok(_) => return Ok(()),
        Err(s) => return Err(s.to_string()),
    }
}

#[tauri::command(rename_all = "snake_case")]
pub fn close_preview(app_manager: State<Mutex<ApplicationManager>>) -> Result<(), String> {
    info!("close_preview command is invoked");
    let result = app_manager.lock().unwrap().close_preview();
    match result {
        Ok(_) => return Ok(()),
        Err(s) => return Err(s.to_string()),
    }
}

File: src/tauri_commands.ts
import { invoke } from "@tauri-apps/api/tauri";

export type tauriTestType = "Mos" | "Thurstone";

export const testTypeToString = (testType: tauriTestType): string => {
  switch (testType) {
    case "Mos":
      return "平均オピニオン評価";
    case "Thurstone":
      return "一対比較法(サーストン法)";
  }
};

//テストマネージャをバックエンドから取得する関数================================
export const tauriGetSettings = (): Promise<{ [key: string]: any }> => {
  return invoke<string[]>("get_settings")
    .then((settings) => {
      let managers: { [key: string]: any } = {};
      for (let setting of settings) {
        let _obj = JSON.parse(setting);
        let _name = _obj["name"];
        managers[_name] = _obj;
      }
      return managers; // `managers` オブジェクトを返す
    })
    .catch((err) => {
      console.error(err);
      throw err; // エラーを再スロー
    });
};

export const tauriAddTest = async (
  testType: tauriTestType,
  jsonString: string
): Promise<void> => {
  return invoke("add_test", { test_type: testType, json_string: jsonString })
    .then(() => {})
    .catch((err) => {
      console.error(err);
      throw err;
    });
};

export const tauriEditTest = async (
  testName: string,
  jsonString: string
): Promise<void> => {
  return invoke("edit_test", { test_name: testName, json_string: jsonString })
    .then(() => {})
    .catch((err) => {
      console.error(err);
      throw err;
    });
};

export const tauriDeleteTest = async (testName: string): Promise<void> => {
  return invoke("delete_test", { test_name: testName })
    .then(() => {})
    .catch((err) => {
      console.error(err);
      throw err;
    });
};

export const tauriStartTest = async (
  testName: string,
  examineeName: string
): Promise<void> => {
  return invoke("start_test", { test_name: testName, examinee: examineeName })
    .then(() => {})
    .catch((err) => {
      console.error(err);
      throw err;
    });
};

export const tauriGetAudio = (): Promise<string[]> => {
  return invoke<string[]>("get_audio")
    .then((paths) => {
      return paths;
    })
    .catch((err) => {
      console.error(err);
      throw err;
    });
};

export const tauriSetScore = (score: string[]): Promise<string> => {
  return invoke<string>("set_score", { score: score })
    .then((status) => {
      return status;
    })
    .catch((err) => {
      console.error(err);
      throw err;
    });
};

export const tauriCloseTest = async (examineeName: string): Promise<void> => {
  return invoke("close_test", { examinee: examineeName })
    .then(() => {})
    .catch((err) => {
      console.error(err);
      throw err;
    });
};

export const tauriDeleteTrial = async (
  testName: string,
  examineeName: string
): Promise<void> => {
  return invoke("delete_trial", { test_name: testName, examinee: examineeName })
    .then(() => {})
    .catch((err) => {
      console.error(err);
      throw err;
    });
};

export const tauriStartPreview = async (testName: string): Promise<void> => {
  return invoke("start_preview", { test_name: testName })
    .then(() => {})
    .catch((err) => {
      console.error(err);
      throw err;
    });
};

export const tauriClosePreview = async (): Promise<void> => {
  return invoke("close_preview")
    .then(() => {})
    .catch((err) => {
      console.error(err);
      throw err;
    });
};
